File: crates/intar-contracts/src/stargate.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

pub const ROUTE_USERNAME_MAX_LEN: usize = 128;
pub const GENERATION_MAX_LEN: usize = 128;

/// Check the public SSH route name. Stargate prints this value, and the public
/// SSH server matches it against the path of an attach call, so it must stay
/// inside a small, non-ambiguous character set.
pub fn validate_route_username(username: &str) -> Result<(), String> {
    if username.is_empty() || username.len() > ROUTE_USERNAME_MAX_LEN {
        return Err(format!(
            "route_username must be 1..={ROUTE_USERNAME_MAX_LEN} characters"
        ));
    }
    if !username
        .bytes()
        .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'))
    {
        return Err(
            "route_username may only contain ASCII letters, digits, '.', '_' and '-'".to_owned(),
        );
    }
    Ok(())
}

#[derive(Clone, Copy, Debug, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum SessionKind {
    NativeSsh,
    BrowserTerminal,
    WorkspaceApp,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TerminalSessionMode {
    Browser,
    Native,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum NativeTerminalAuthMode {
    ProfileKeys,
}

/// The complete SSH endpoint of one scenario VM terminal. Only the admin API
/// carries this value: `private_key_openssh` is the credential that reaches
/// the guest, and the gateway never sends it to a browser or a native client.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct TerminalTarget {
    pub username: String,
    pub transport: SshTargetTransport,
    pub host_key_openssh: String,
    pub private_key_openssh: String,
    pub authorized_client_public_keys_openssh: Vec<String>,
}

/// One terminal route. A browser route starts pending: the route exists and a
/// browser socket can connect, but Stargate must not dial the guest yet. An
/// admin attach moves the route to ready, and only then does the PTY and the
/// recording start.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(tag = "state", rename_all = "snake_case")]
pub enum TerminalTargetState {
    Pending,
    Ready(Box<TerminalTarget>),
}

impl TerminalTargetState {
    pub fn ready_target(&self) -> Option<&TerminalTarget> {
        match self {
            Self::Pending => None,
            Self::Ready(target) => Some(target),
        }
    }
}

/// Route identity. Every field is mandatory and non-empty: a pending route
/// must name the host, the run, the VM, and the user that it belongs to.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct RouteMetadata {
    pub host_id: String,
    pub run_id: String,
    pub vm_id: String,
    pub user_id: String,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct IssueTerminalSessionRequest {
    pub route_username: String,
    /// Opaque route generation. An attach call must repeat this exact string.
    pub generation: String,
    /// `pending` for a browser route. `ready` with the complete endpoint for a
    /// native route.
    pub target: TerminalTargetState,
    pub route_expires_at: i64,
    pub mode: TerminalSessionMode,
    pub metadata: RouteMetadata,
}

/// Stage a ready target on a pending browser route. The gateway stores the
/// target and returns an attachment identifier, but the route is still not
/// ready: no waiter wakes and no socket dials. The control plane activates the
/// attachment after its own admission fence.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct StageTerminalTargetRequest {
    pub run_id: String,
    pub vm_id: String,
    pub user_id: String,
    pub generation: String,
    /// The complete target. There is no pending shape on a stage call: a call
    /// that carries no target is a validation error.
    pub target: TerminalTarget,
}

/// The identifier of one staged target. An identical repeat of a stage call
/// returns the same value, so a lost answer is safe to retry.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct StageTerminalTargetResponse {
    pub attachment_id: String,
}

/// Activate one staged target. Only this call makes the route ready and wakes
/// the waiting browser socket.
#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct ActivateTerminalTargetRequest {
    pub run_id: String,
    pub vm_id: String,
    pub user_id: String,
    pub generation: String,
    pub attachment_id: String,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct BrowserTerminalSession {
    pub websocket_url: String,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct NativeTerminalSession {
    pub auth_mode: NativeTerminalAuthMode,
    pub authorized_key_count: usize,
    pub ssh_host: String,
    pub ssh_port: u16,
    pub username: String,
    pub public_host_key_openssh: String,
    pub public_host_key_fingerprint_sha256: String,
    pub known_hosts_line: String,
    pub command: String,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct IssueTerminalSessionResponse {
    pub route_username: String,
    pub expires_at: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub browser: Option<BrowserTerminalSession>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub native: Option<NativeTerminalSession>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkspaceAppProtocol {
    Http,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct IssueWorkspaceAppSessionRequest {
    /// Random, human-readable identifier generated by the control plane. It
    /// is a public route identity, not an authorization secret, and must not
    /// contain user-provided labels or resource names.
    pub route_id: String,
    /// Refuse to replace an existing route with this identifier.
    #[serde(default)]
    pub create_only: bool,
    pub target_username: String,
    /// Literal IP and port of the guest SSH endpoint already exposed to
    /// Stargate by the VM harness.
    pub transport: SshTargetTransport,
    pub target_host_key_openssh: String,
    pub target_private_key_openssh: String,
    /// Allowlisted application port inside the guest. Stargate reaches it
    /// through an SSH direct-tcpip channel; it is never published on the host.
    pub target_app_port: u16,
    pub protocol: WorkspaceAppProtocol,
    /// Optional virtual-host value sent only to the declared guest service.
    /// The public route hostname remains available as X-Forwarded-Host.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub upstream_host: Option<String>,
    pub route_expires_at: i64,
    #[serde(default)]
    pub metadata: WorkspaceAppMetadata,
}

/// Workspace app routes keep optional metadata: they carry no terminal route
/// identity, and an absent value is not an authorization decision.
#[derive(Clone, Debug, Default, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct WorkspaceAppMetadata {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub host_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub run_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub vm_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub struct IssueWorkspaceAppSessionResponse {
    pub route_id: String,
    /// URL containing a short-lived, single-use bootstrap capability. The
    /// public gateway exchanges it for a route-bound browser session and
    /// redirects to the same URL without the capability.
    pub url: String,
    pub bootstrap_expires_at: i64,
    pub expires_at: i64,
}

#[derive(Clone, Debug, Deserialize, Eq, Hash, JsonSchema, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct HostRelayIdentity {
    pub host_id: String,
    pub session_id: String,
    pub credential_generation: u64,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, JsonSchema, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RelayService {
    Ssh,
}

#[derive(Clone, Debug, Deserialize, Eq, Hash, JsonSchema, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct RelayTarget {
    pub host: HostRelayIdentity,
    pub owner_id: String,
    pub execution_id: String,
    #[schemars(range(min = 1))]
    pub execution_generation: u64,
    pub vm_id: String,
    pub service: RelayService,
}

#[derive(Clone, Debug, Deserialize, Eq, Hash, JsonSchema, PartialEq, Serialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
pub enum SshTargetTransport {
    Direct { host: String, port: u16 },
    Relay { target: RelayTarget },
}

impl HostRelayIdentity {
    pub fn validate(&self) -> std::io::Result<()> {
        relay_check(
            valid_relay_id(&self.host_id)
                && valid_relay_id(&self.session_id)
                && self.credential_generation > 0,
            "invalid relay host identity",
        )
    }
}
impl RelayTarget {
    pub fn validate(&self) -> std::io::Result<()> {
        self.host.validate()?;
        relay_check(
            [&self.owner_id, &self.execution_id, &self.vm_id]
                .into_iter()
                .all(|id| valid_relay_id(id))
                && self.execution_generation > 0,
            "invalid relay workload identity",
        )
    }
}
impl SshTargetTransport {
    pub fn validate(&self) -> std::io::Result<()> {
        match self {
            Self::Direct { host, port } => relay_check(
                host.parse::<std::net::IpAddr>().is_ok() && *port > 0,
                "direct SSH target requires a literal IP and nonzero port",
            ),
            Self::Relay { target } => target.validate(),
        }
    }
}
fn relay_check(valid: bool, message: &str) -> std::io::Result<()> {
    if valid {
        Ok(())
    } else {
        Err(std::io::Error::new(
            std::io::ErrorKind::InvalidInput,
            message,
        ))
    }
}
fn valid_relay_id(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 128
        && id
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'-' | b'_' | b'.'))
}

File: apps/web/src/lib/stargate.ts
import { env } from "cloudflare:workers";
import { createAppId } from "@/lib/id";
import type {
  ActivateTerminalTargetRequest as StargateActivateRequest,
  IssueTerminalSessionResponse as StargateApiTerminalSessionResponse,
  IssueTerminalSessionRequest as StargateCreateRequest,
  RouteMetadata as StargateRouteMetadata,
  StageTerminalTargetRequest as StargateStageRequest,
  StageTerminalTargetResponse as StargateStageResponse,
  TerminalTarget as StargateWireTerminalTarget,
  TerminalTargetState as StargateWireTargetState,
} from "@/generated/stargate";
import {
  buildTemporaryNativeSshCommand,
  temporaryNativeSshKeyFilename,
} from "@/lib/native-ssh";

const DEFAULT_ASSERTION_HEADER = "cf-access-jwt-assertion";
const DEFAULT_ROUTE_TTL_SECONDS = 4 * 60 * 60;
const DEFAULT_STARGATE_ADMIN_ORIGIN = "http://stargate.internal";
const STARGATE_ADMIN_CREATE_TIMEOUT_MS = 30_000;
const textEncoder = new TextEncoder();

interface AssertionTokenOptions {
  secret: string;
  issuer: string;
  audience: string;
  subject: string;
  ttlSeconds: number;
}

/** The four mandatory non-empty route identity fields. */
function routeMetadata(metadata: {
  hostId: string;
  runId: string;
  vmId: string;
  userId: string;
}): StargateRouteMetadata {
  return {
    host_id: metadata.hostId,
    run_id: metadata.runId,
    vm_id: metadata.vmId,
    user_id: metadata.userId,
  };
}

/** The SSH endpoint of one guest. A browser route receives this after the
 * gateway attaches the ready target; it never travels through the browser. */
export interface StargateTerminalTarget {
  username: string;
  host: string;
  port: number;
  hostKeyOpenssh: string;
  privateKeyOpenssh: string;
  authorizedClientPublicKeysOpenssh: string[];
}

/**
 * The web app works in camelCase; the wire uses the generated snake_case
 * names. The return type is the generated contract type, so a field that Rust
 * adds, renames, or removes makes this function fail to compile instead of
 * silently sending a stale body.
 */
function wireTerminalTarget(
  target: StargateTerminalTarget,
): StargateWireTerminalTarget {
  return {
    username: target.username,
    host: target.host,
    port: target.port,
    host_key_openssh: target.hostKeyOpenssh,
    private_key_openssh: target.privateKeyOpenssh,
    authorized_client_public_keys_openssh:
      target.authorizedClientPublicKeysOpenssh,
  };
}

interface IssueTerminalSessionCommonInput {
  routeUsername: string;
  /**
   * Opaque route generation. The gateway compares it as an exact string on
   * the attach call, so a revoked route can not be revived by a stale attach.
   */
  generation: string;
  expiresAt: Date;
  metadata: {
    hostId: string;
    runId: string;
    vmId: string;
    userId: string;
  };
}

/**
 * A browser route starts with no target and waits for the admin attach. A
 * native route starts with a ready target in the create call.
 */
export type IssueStargateTerminalSessionInput =
  | (IssueTerminalSessionCommonInput & { mode: "browser" })
  | (IssueTerminalSessionCommonInput & {
      mode: "native";
      target: StargateTerminalTarget;
      temporaryClientPublicKeyOpenssh?: string;
    });

export interface BrowserTerminalSessionResult {
  routeUsername: string;
  expiresAt: number;
  generation: string;
  browser: {
    websocketUrl: string;
  };
  native?: undefined;
}

export interface NativeTerminalSessionResult {
  routeUsername: string;
  expiresAt: number;
  generation: string;
  native: {
    authMode: "profile_keys" | "issued_key";
    authorizedKeyCount: number;
    host: string;
    port: number;
    username: string;
    publicHostKeyOpenssh: string;
    publicHostKeyFingerprintSha256: string;
    knownHostsLine: string;
    command: string;
    keyFilename?: string;
  };
  browser?: undefined;
}

export type StargateTerminalSessionResult =
  | BrowserTerminalSessionResult
  | NativeTerminalSessionResult;

export async function issueStargateTerminalSession(
  input: IssueStargateTerminalSessionInput,
): Promise<StargateTerminalSessionResult> {
  const generation = requiredValue(input.generation, "generation");
  const temporaryClientPublicKeyOpenssh =
    input.mode === "native"
      ? input.temporaryClientPublicKeyOpenssh?.trim()
      : undefined;
  if (
    input.mode === "native" &&
    input.temporaryClientPublicKeyOpenssh !== undefined &&
    !temporaryClientPublicKeyOpenssh
  ) {
    throw new Error("temporary native SSH public key is empty");
  }
  if (
    temporaryClientPublicKeyOpenssh &&
    input.mode === "native" &&
    input.target.authorizedClientPublicKeysOpenssh.length > 0
  ) {
    throw new Error(
      "temporary native SSH authorization cannot be combined with another route mode or profile key",
    );
  }
  const authorizedClientPublicKeysOpenssh =
    input.mode === "native" && temporaryClientPublicKeyOpenssh !== undefined
      ? [temporaryClientPublicKeyOpenssh]
      : input.mode === "native"
        ? input.target.authorizedClientPublicKeysOpenssh
        : [];
  const targetBody: StargateWireTargetState =
    input.mode === "native"
      ? {
          state: "ready",
          ...wireTerminalTarget(input.target),
          // The temporary key replaces the profile-key list for this route.
          authorized_client_public_keys_openssh:
            authorizedClientPublicKeysOpenssh,
        }
      : { state: "pending" };
  const createBody: StargateCreateRequest = {
    route_username: input.routeUsername,
    generation,
    target: targetBody,
    route_expires_at: Math.floor(input.expiresAt.getTime() / 1000),
    mode: input.mode,
    metadata: routeMetadata(input.metadata),
  };
  const response = await stargateAdminFetch("/v1/terminal-sessions", {
    method: "POST",
    signal: AbortSignal.timeout(STARGATE_ADMIN_CREATE_TIMEOUT_MS),
    headers: {
      "content-type": "application/json",
      [assertionHeader(env.STARGATE_ADMIN_AUTH_HEADER)]:
        await createAssertionToken({
          secret: requiredValue(
            env.STARGATE_ADMIN_AUTH_SECRET,
            "STARGATE_ADMIN_AUTH_SECRET",
          ),
          issuer: requiredValue(
            env.STARGATE_ADMIN_AUTH_ISSUER,
            "STARGATE_ADMIN_AUTH_ISSUER",
          ),
          audience: requiredValue(
            env.STARGATE_ADMIN_AUTH_AUDIENCE,
            "STARGATE_ADMIN_AUTH_AUDIENCE",
          ),
          subject: "intar-admin",
          ttlSeconds: 60,
        }),
    },
    body: JSON.stringify(createBody),
  });

  if (!response.ok) {
    throw new Error(
      `stargate terminal session create failed (${response.status})`,
    );
  }

  const body =
    (await response.json()) as Partial<StargateApiTerminalSessionResponse>;

  if (
    typeof body.route_username !== "string" ||
    typeof body.expires_at !== "number"
  ) {
    throw new Error("invalid stargate terminal session response");
  }

  if (input.mode === "browser") {
    if (typeof body.browser?.websocket_url !== "string") {
      throw new Error("invalid stargate browser terminal response");
    }
    return {
      routeUsername: body.route_username,
      expiresAt: body.expires_at * 1000,
      generation,
      browser: {
        websocketUrl: body.browser.websocket_url,
      },
    };
  }

  if (
    body.native?.auth_mode !== "profile_keys" ||
    typeof body.native?.authorized_key_count !== "number" ||
    typeof body.native?.ssh_host !== "string" ||
    typeof body.native?.ssh_port !== "number" ||
    typeof body.native?.username !== "string" ||
    typeof body.native?.public_host_key_openssh !== "string" ||
    typeof body.native?.public_host_key_fingerprint_sha256 !== "string" ||
    typeof body.native?.known_hosts_line !== "string" ||
    typeof body.native?.command !== "string"
  ) {
    throw new Error("invalid stargate native terminal response");
  }
  const keyFilename = temporaryClientPublicKeyOpenssh
    ? temporaryNativeSshKeyFilename(body.route_username)
    : undefined;
  const native = {
    authMode: input.temporaryClientPublicKeyOpenssh
      ? ("issued_key" as const)
      : body.native.auth_mode,
    authorizedKeyCount: body.native.authorized_key_count,
    host: body.native.ssh_host,
    port: body.native.ssh_port,
    username: body.native.username,
    publicHostKeyOpenssh: body.native.public_host_key_openssh,
    publicHostKeyFingerprintSha256:
      body.native.public_host_key_fingerprint_sha256,
    knownHostsLine: body.native.known_hosts_line,
    command: keyFilename
      ? buildTemporaryNativeSshCommand({
          username: body.native.username,
          host: body.native.ssh_host,
          port: body.native.ssh_port,
          knownHostsLine: body.native.known_hosts_line,
          keyFilename,
        })
      : body.native.command,
    ...(keyFilename ? { keyFilename } : {}),
  };
  return {
    routeUsername: body.route_username,
    expiresAt: body.expires_at * 1000,
    generation,
    native,
  };
}

/**
 * Phase one. Stores the ready target on a pending browser route and returns
 * the attachment id, but does NOT activate it: the gateway starts no PTY and
 * wakes no waiting socket until the activate call.
 *
 * The call is idempotent for one generation: a repeat stage for the same
 * ready target returns the same attachment id, so a retry after a lost
 * response is safe.
 */
export async function stageStargateTerminalTarget(input: {
  routeUsername: string;
  generation: string;
  runId: string;
  vmId: string;
  userId: string;
  target: StargateTerminalTarget;
}): Promise<{ attachmentId: string }> {
  const response = await stargateAdminFetch(
    `/v1/terminal-sessions/${encodeURIComponent(requiredValue(input.routeUsername, "routeUsername"))}/target`,
    {
      method: "POST",
      signal: AbortSignal.timeout(STARGATE_ADMIN_CREATE_TIMEOUT_MS),
      headers: {
        "content-type": "application/json",
        [assertionHeader(env.STARGATE_ADMIN_AUTH_HEADER)]:
          await createAssertionToken({
            secret: requiredValue(
              env.STARGATE_ADMIN_AUTH_SECRET,
              "STARGATE_ADMIN_AUTH_SECRET",
            ),
            issuer: requiredValue(
              env.STARGATE_ADMIN_AUTH_ISSUER,
              "STARGATE_ADMIN_AUTH_ISSUER",
            ),
            audience: requiredValue(
              env.STARGATE_ADMIN_AUTH_AUDIENCE,
              "STARGATE_ADMIN_AUTH_AUDIENCE",
            ),
            subject: "intar-admin",
            ttlSeconds: 60,
          }),
      },
      body: JSON.stringify({
        run_id: input.runId,
        vm_id: input.vmId,
        user_id: input.userId,
        generation: requiredValue(input.generation, "generation"),
        // A stage carries the complete target. There is no pending shape on a
        // stage call, so the target has no `state` tag.
        target: wireTerminalTarget(input.target),
      } satisfies StargateStageRequest),
    },
  );

  // 404 and 409 are terminal for the route: it is gone, expired, or bound to
  // other metadata. Any other failure is ambiguous and reaches the caller,
  // which fails closed.
  if (!response.ok) {
    throw new StargateTerminalAttachError(response.status);
  }
  // Typed against the generated envelope so a renamed field fails to compile
  // instead of silently reading undefined.
  const body = (await response.json().catch(() => null)) as
    | Partial<StargateStageResponse>
    | null;
  const attachmentId = body?.attachment_id;
  if (typeof attachmentId !== "string" || !attachmentId) {
    // A staged target without an id can not be activated, so the caller must
    // treat this as an ambiguous failure and fail closed.
    throw new StargateTerminalAttachError(response.status);
  }
  return { attachmentId };
}

/**
 * Phase two. Activates one staged attachment. The gateway opens the PTY and
 * wakes the waiting socket only here, so a staged attachment that is never
 * activated can not produce a usable shell.
 *
 * Activation is idempotent for one generation and attachment, so a retry after
 * a lost response can not open a second shell.
 */
export async function activateStargateTerminalTarget(input: {
  routeUsername: string;
  generation: string;
  attachmentId: string;
  runId: string;
  vmId: string;
  userId: string;
}): Promise<void> {
  const response = await stargateAdminFetch(
    `/v1/terminal-sessions/${encodeURIComponent(requiredValue(input.routeUsername, "routeUsername"))}/activate`,
    {
      method: "POST",
      signal: AbortSignal.timeout(STARGATE_ADMIN_CREATE_TIMEOUT_MS),
      headers: {
        "content-type": "application/json",
        [assertionHeader(env.STARGATE_ADMIN_AUTH_HEADER)]:
          await createAssertionToken({
            secret: requiredValue(
              env.STARGATE_ADMIN_AUTH_SECRET,
              "STARGATE_ADMIN_AUTH_SECRET",
            ),
            issuer: requiredValue(
              env.STARGATE_ADMIN_AUTH_ISSUER,
              "STARGATE_ADMIN_AUTH_ISSUER",
            ),
            audience: requiredValue(
              env.STARGATE_ADMIN_AUTH_AUDIENCE,
              "STARGATE_ADMIN_AUTH_AUDIENCE",
            ),
            subject: "intar-admin",
            ttlSeconds: 60,
          }),
      },
      body: JSON.stringify({
        run_id: input.runId,
        vm_id: input.vmId,
        user_id: input.userId,
        generation: requiredValue(input.generation, "generation"),
        attachment_id: requiredValue(input.attachmentId, "attachmentId"),
      } satisfies StargateActivateRequest),
    },
  );

  // Every non-success outcome is ambiguous for a call that opens a shell, so
  // it reaches the caller, which revokes this generation and fails closed.
  if (response.status === 204 || response.status === 200) {
    return;
  }
  throw new StargateTerminalAttachError(response.status);
}

export class StargateTerminalAttachError extends Error {
  constructor(readonly status: number) {
    super(`stargate terminal target attach failed (${status})`);
    this.name = "StargateTerminalAttachError";
  }
}


export async function deleteStargateTerminalRoute(
  routeUsername: string,
  generation: string,
): Promise<void> {
  const trimmed = routeUsername.trim();
  if (!trimmed) {
    return;
  }
  const routeGeneration = requiredValue(generation, "generation");

  const response = await stargateAdminFetch(
    `/v1/routes/${encodeURIComponent(trimmed)}?generation=` +
      encodeURIComponent(routeGeneration),
    {
      method: "DELETE",
      headers: {
        [assertionHeader(env.STARGATE_ADMIN_AUTH_HEADER)]:
          await createAssertionToken({
            secret: requiredValue(
              env.STARGATE_ADMIN_AUTH_SECRET,
              "STARGATE_ADMIN_AUTH_SECRET",
            ),
            issuer: requiredValue(
              env.STARGATE_ADMIN_AUTH_ISSUER,
              "STARGATE_ADMIN_AUTH_ISSUER",
            ),
            audience: requiredValue(
              env.STARGATE_ADMIN_AUTH_AUDIENCE,
              "STARGATE_ADMIN_AUTH_AUDIENCE",
            ),
            subject: "intar-admin",
            ttlSeconds: 60,
          }),
      },
    },
  );

  // 204 deleted, 404 already gone, 409 still owned by a newer generation.
  // Every other status is a real failure and must reach the caller.
  if (
    response.status === 204 ||
    response.status === 404 ||
    response.status === 409
  ) {
    return;
  }
  throw new Error(`stargate route delete failed (${response.status})`);
}

export async function deleteStargateRoute(
  routeUsername: string,
): Promise<void> {
  const trimmed = routeUsername.trim();
  if (!trimmed) {
    return;
  }

  const response = await stargateAdminFetch(
    `/v1/routes/${encodeURIComponent(trimmed)}`,
    {
      method: "DELETE",
      headers: {
        [assertionHeader(env.STARGATE_ADMIN_AUTH_HEADER)]:
          await createAssertionToken({
            secret: requiredValue(
              env.STARGATE_ADMIN_AUTH_SECRET,
              "STARGATE_ADMIN_AUTH_SECRET",
            ),
            issuer: requiredValue(
              env.STARGATE_ADMIN_AUTH_ISSUER,
              "STARGATE_ADMIN_AUTH_ISSUER",
            ),
            audience: requiredValue(
              env.STARGATE_ADMIN_AUTH_AUDIENCE,
              "STARGATE_ADMIN_AUTH_AUDIENCE",
            ),
            subject: "intar-admin",
            ttlSeconds: 60,
          }),
      },
    },
  );

  if (response.status === 404 || response.status === 204) {
    return;
  }
  if (!response.ok) {
    throw new Error(`stargate route delete failed (${response.status})`);
  }
}

export function stargateRouteTtlMs(): number {
  const raw = env.STARGATE_ROUTE_TTL_SECONDS?.trim();
  const ttlSeconds = raw ? Number.parseInt(raw, 10) : DEFAULT_ROUTE_TTL_SECONDS;
  const normalized =
    Number.isFinite(ttlSeconds) && ttlSeconds > 0
      ? ttlSeconds
      : DEFAULT_ROUTE_TTL_SECONDS;
  return normalized * 1000;
}

async function stargateAdminFetch(
  path: string,
  init: RequestInit,
): Promise<Response> {
  const binding = env.STARGATE_ADMIN_SERVICE;
  if (binding) {
    return binding.fetch(new URL(path, DEFAULT_STARGATE_ADMIN_ORIGIN), init);
  }

  const baseUrl = env.STARGATE_ADMIN_BASE_URL?.trim();
  if (baseUrl) {
    return fetch(new URL(path, requiredUrl(baseUrl)), init);
  }

  throw new Error(
    "STARGATE_ADMIN_SERVICE binding or STARGATE_ADMIN_BASE_URL is required",
  );
}

async function createAssertionToken(
  options: AssertionTokenOptions,
): Promise<string> {
  const nowSeconds = Math.floor(Date.now() / 1000);
  const payload = {
    iss: options.issuer,
    aud: options.audience,
    sub: options.subject,
    iat: nowSeconds,
    nbf: nowSeconds,
    exp: nowSeconds + options.ttlSeconds,
    jti: createAppId(),
  };

  const encodedHeader = base64UrlEncode(
    textEncoder.encode(JSON.stringify({ alg: "HS256", typ: "JWT" })),
  );
  const encodedPayload = base64UrlEncode(
    textEncoder.encode(JSON.stringify(payload)),
  );
  const signingInput = `${encodedHeader}.${encodedPayload}`;

  const key = await crypto.subtle.importKey(
    "raw",
    textEncoder.encode(options.secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const signature = await crypto.subtle.sign(
    "HMAC",
    key,
    textEncoder.encode(signingInput),
  );

  return `${signingInput}.${base64UrlEncode(new Uint8Array(signature))}`;
}

function base64UrlEncode(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/g, "");
}

function assertionHeader(value?: string): string {
  const trimmed = value?.trim();
  return trimmed || DEFAULT_ASSERTION_HEADER;
}

function requiredValue(value: string | undefined, name: string): string {
  const trimmed = value?.trim();
  if (!trimmed) {
    throw new Error(`${name} is required`);
  }
  return trimmed;
}

function requiredUrl(value: string | undefined): URL {
  return new URL(requiredValue(value, "stargate base url"));
}
